### DISCUSSION TITLE
Having an US Transducer change orientation with GUI checkboxes

### DISCUSSION THREAD

**Discussion by nathanbmnt** [PlusToolkit/PlusLib#964] [Q&A]
Hello,
I have an ultrasound image streaming to 3D Slicer through PlusServer. The US transducer can be facing up, down, or flipped LR. There is no encoder/camera, so I have a checkbox in the Slicer GUI to apply an up-down flip, and another checkbox for applying a left-right flip to the image.
![image](https://user-images.githubusercontent.com/40271210/176478151-359e7568-7d92-46d7-81a8-b7cb8f03dd33.png)

Users can click an acquire button which saves a single frame .mhd via VirtualCapture.

Users can then convert the mhd to mha with VolumeReconstructor.exe, and load the mha into 3D Slicer. I want the volume in 3D slicer to have the flips burned in.

What is the best way to go about burning the flips in?

I can think of a few options:
1. Create a new class `vtkPlusManualTracker` and use IGTLinkRemote commands from 3D Slicer to change the us probe transform.
2. Save the mha with the wrong orientation but write a custom header like `is_flipped_lr`. When VolumeReconstructor.exe is run through python change the `--image-to-reference-transform` to include the flips.
3. Use SetInputImageOrientation on the vtkPlusDataSource to change MF->UF or MN or UN

**Comment by ungi** [comment]
Hi, there are two different things here, display and volume reconstruction.
For display, there is a module in SlicerIGT, called Volume Reslice Driver. You can use that to make any 2D view "follow" the ultrasound image. In the advanced options, you can flip and rotate the image any way. (There is only one flip option, but flip+180 rotation is the other flip.)
For volume reconstruction, you need to set up your PLUS config file correctly and never change it to flip the image. The PortUsImageOrientation parameter in your PLUS config file needs to match the orientation of the images received from the ultrasound machine. If you set up that parameter correctly, PLUS will keep track of the correct orientation, and it will run volume reconstruction correctly.
It's not clear to me from your message if you have a position tracker on your ultrasound or not. You can only reconstruct 3D volumes from ultrasound sequences if you have a tracker attached and calibrated on your ultrasound transducer.

**Comment by nathanbmnt** [comment]
> For display, there is a module in SlicerIGT, called Volume Reslice Driver

I've been using a vtkMRMLLinearTransformNode, but I'll look into this it may be cleaner.


> It's not clear to me from your message if you have a position tracker on your ultrasound or not

There is no position tracker on the hardware. What a user does is rotate the transducer and then click either of the two check boxes in the GUI.

> You can only reconstruct 3D volumes from ultrasound sequences

Ultimately after the "acquire" button is clicked we want a .mha saved in the correct orientation. Our current approach is to save a single frame mhd sequence and then use VolumeReconstructor to convert it to .mha.

So for an mhd sequence we want it to look like this if there was an up-down flip:
![image](https://user-images.githubusercontent.com/40271210/176715774-b81c0149-d4c7-4043-b412-e42094d8ac6b.png)

**Comment by ungi** [comment]
VolumeReconstructor is for reconstructing 3D volumes from tracked ultrasound images. If the recorded 2D images don't have tracking information saved by a tracker device, then I don't think you can reconstruct a volume.

**Comment by nathanbmnt** [comment]
Does Plus have another way of saving a 2D image to disk that has an orientation in 3D space when loaded into 3D Slicer?

That's why I'm thinking of creating a new device class called  `vtkPlusManualTracker` where the transforms are changed infrequently and you set the transforms of the tracker with IGTLinkRemote.

**Comment by nathanbmnt** [comment]
@Sunderlandkyl @lassoan If I made a `vtkPlusManualTracker` class where the transform is changed with PlusRemote commands would this be acceptable in the PlusLib repo?

**Reply by Sunderlandkyl** [reply]
You could use the existing [OpenIGTLinkTracker](http://perk-software.cs.queensu.ca/plus/doc/nightly/user/DeviceOpenIGTLinkTracker.html) device and send the transforms to Plus that way.

**Reply by nathanbmnt** [reply]
Thank you for the help, I think [OpenIGTLinkTracker](http://perk-software.cs.queensu.ca/plus/doc/nightly/user/DeviceOpenIGTLinkTracker.html) is what we need. But how would I go about sending a message from 3D Slicer with Python?

For a GenericSerialDevice you can send commands using the vtkMRMLIGTLConnectorNode which is a client connected to the port in the `<PlusOpenIGTLinkServer>` item in the xml.
http://perk-software.cs.queensu.ca/plus/doc/nightly/user/DeviceGenericSerial.html

However in OpenIGTLinkTracker you specify `ServerPort` and `ServerAddress` in the `<Device>` settings? And not in `<PlusOpenIGTLinkServer>` settings?

**Reply by Sunderlandkyl** [reply]
If you create an OpenIGTLinkIF connector node in Slicer with the "Server" type, you can set the outgoing nodes from Slicer.
As the nodes are modified, it will send the updates to any connected clients, including Plus. Just make sure that you use different ports for the incoming/outgoing connections.

**Reply by Sunderlandkyl** [reply]
Note that you can also do volume reconstruction on the Slicer side using the "Volume Reconstruction" module in SlicerIGT. The module uses the same algorithm as Plus to perform the volume reconstruction.

**Reply by nathanbmnt** [reply]
Thanks I will try this and mark this discussion as answered if I see the transform is being updated in my PlusServer VirtualDevice recording.

**Reply by nathanbmnt** [reply]
I think OpenIGTLinkTracker is working well but after connecting to the VirtualMixer and changing the transform I get this error when I change the transform:

`|ERROR|054.097000| TrackerDevice-USToReference: vtkPlusBuffer: Cannot perform interpolation, time difference compared to itemB is too big 0.868929 ( itemBtime: 53.194000, requested time: 54.062929).| in E:\D\PSNP64b\PlusLib\src\PlusDataCollection\vtkPlusBuffer.cxx(1183)`

 Is there some interpolation setting I have to turn off? I have `UseReceivedTimestamps ="FALSE"` and `UseLastTransformsOnReceiveTimeout ="TRUE"`.

**Reply by nathanbmnt** [reply]
I believe this is a bug with how the VirtualMixer and OpenIGTLinkTracker are interacting so I've created an issue here:
https://github.com/PlusToolkit/PlusLib/issues/967
